### PR TITLE
Add optional parameters

### PR DESCRIPTION
**Pull request by int-index**

### CODE REVIEW

File: src/Named/Internal.hs [hunk 1]
```diff
@@ -1,59 +1,129 @@
...
+
+#if MIN_VERSION_base(4,10,0)
+{-# COMPLETE Arg #-}
+#endif
+
+-- | Infix notation for the type of a named parameter.
+type name :! a = NamedF Identity a name
+
+-- | Infix notation for the type of an optional named parameter.
+type name :? a = NamedF Maybe a name
```
Comment (ChShersh) [inline src/Named/Internal.hs]: Wow, I like this generalized solution and cool `:?` for optional arguments. This design looks pretty good.

File: src/Named/Internal.hs [hunk 2]
```diff
@@ -1,59 +1,129 @@
...
+  ArgF (f a) -- ^ Match on an F-argument without specifying its name.
+             -- See also: 'argF'.
 
-instance (name ~ name', a ~ a') => IsLabel name (a -> Named a' name') where
+-- | Match on an argument without specifying its name. See also: 'arg'.
+pattern Arg :: a -> name :! a
+pattern Arg a = ArgF (Identity a)
+
+#if MIN_VERSION_base(4,10,0)
+{-# COMPLETE Arg #-}
```
Comment (ChShersh) [inline src/Named/Internal.hs]: Why do you need COMPLETE pragma here? `ArgF` is the only constructor of `NamedF` and `Identity` is the only constructor of `Identity`. It doesn't make sense to me why `ghc` could complain on incomplete patterns...

File: src/Named/Internal.hs [hunk 3]
```diff
@@ -1,59 +1,129 @@
...
+  {-# INLINE with #-}
+
+data Defaults = Defaults
+
+{- |
+Passing 'defaults' to a function fills all unspecified optional parameters
+with 'Nothing':
+
+@
+fn            :: "b" ':!' Bool -> "x" ':?' Char -> Int -> IO ()
```
Comment (ChShersh) [inline src/Named/Internal.hs]: This alignment also looks not perfect

File: src/Named/Internal.hs [hunk 4]
```diff
@@ -102,71 +158,88 @@ signature for @fn@ is required. In case a type signature for @fn@ is
...
+data DFill
+data DPass
 
-type family Decide (name :: Symbol) (fn :: Type) :: Decision where
-  Decide name (Named a name -> r) = Done
-  Decide name (x -> r) = Skip (Decide name r)
-  Decide name t =
+type family Decide (p :: Type) (fn :: Type) :: [Type] where
+  Decide (NamedF f' a' name) (NamedF f a name -> r) = '[DApply]
+  Decide Defaults (NamedF Maybe a name -> r) = DFill : Decide Defaults r
```
Comment (ChShersh) [inline src/Named/Internal.hs]: I think it's okay to have special case for `Maybe`. Once you discover use cases for other types like `NonEmpty` you just need to add extra case here (in perfect case) and invent new `:!`-like operator.

File: src/Named.hs [hunk 5]
```diff
@@ -85,25 +94,31 @@ the function and supply the first keyword argument with a matching name.
 For example:
 
 @
-bar           :: A \``Named`\` "x" -> B \``Named`\` "y" -> IO ()
-bar '!' #y b    :: A \``Named`\` "x"                  -> IO ()
-'with' #y b bar :: A \``Named`\` "x"                  -> IO ()
+bar             :: "x" :! A -> "y" :! B -> IO ()
```
[the remark this answers is not in the document]
Comment (int-index) [reply src/Named.hs]: Are you reading the source or the rendered docs?

File: src/Named/Internal.hs (hunk 2 above)
Comment (int-index) [reply src/Named/Internal.hs]: GHC has issues with exhaustiveness checking of pattern synonyms.

**Pull request merged**